src/torcontrol.cpp

```diff
@@ -0,0 +1,685 @@
+    std::vector<uint8_t> computedHash(CHMAC_SHA256::OUTPUT_SIZE, 0);
+    computeHash.Write(begin_ptr(cookie), cookie.size());
+    computeHash.Write(begin_ptr(clientNonce), clientNonce.size());
+    computeHash.Write(begin_ptr(serverNonce), serverNonce.size());
+    computeHash.Finalize(begin_ptr(computedHash));
+    return computedHash;
+}
+
+void TorController::authchallenge_cb(TorControlConnection& conn, const TorControlReply& reply)
+{
+    if (reply.code == 250) {
+        LogPrint("tor", "tor: SAFECOOKIE authentication challenge succesful\n");
+        std::pair<std::string,std::string> l = SplitTorReplyLine(reply.lines[0]);
+        if (l.first == "AUTHCHALLENGE") {
+            std::map<std::string,std::string> m = ParseTorReplyMapping(l.second);
+            std::vector<uint8_t> serverHash = ParseHex(m["SERVERHASH"]);
+            std::vector<uint8_t> serverNonce = ParseHex(m["SERVERNONCE"]);
+            LogPrint("tor", "tor: AUTHCHALLENGE ServerHash %s ServerNonce %s\n", HexStr(serverHash), HexStr(serverNonce));
+            if (serverNonce.size() != 32) {
+                LogPrintf("tor: ServerNonce is not 32 bytes, as required by spec\n");
+                return;
+            }
+
+            std::vector<uint8_t> computedServerHash = ComputeResponse(TOR_SAFE_SERVERKEY, cookie, clientNonce, serverNonce);
+            if (computedServerHash != serverHash) {
+                LogPrintf("tor: ServerHash %s does not match expected ServerHash %s\n", HexStr(serverHash), HexStr(computedServerHash));
+                return;
+            }
+
+            std::vector<uint8_t> computedClientHash = ComputeResponse(TOR_SAFE_CLIENTKEY, cookie, clientNonce, serverNonce);
+            conn.Command("AUTHENTICATE " + HexStr(computedClientHash), boost::bind(&TorController::auth_cb, this, _1, _2));
+        } else {
+            LogPrintf("tor: Invalid reply to AUTHCHALLENGE\n");
+        }
+    } else {
+        LogPrintf("tor: SAFECOOKIE authentication challenge failed\n");
+    }
+}
+
+void TorController::protocolinfo_cb(TorControlConnection& conn, const TorControlReply& reply)
+{
+    if (reply.code == 250) {
+        std::set<std::string> methods;
+        std::string cookiefile;
+        /*
+         * 250-AUTH METHODS=COOKIE,SAFECOOKIE COOKIEFILE="/home/x/.tor/control_auth_cookie"
+         * 250-AUTH METHODS=NULL
+         * 250-AUTH METHODS=HASHEDPASSWORD
+         */
+        BOOST_FOREACH(const std::string &s, reply.lines) {
+            std::pair<std::string,std::string> l = SplitTorReplyLine(s);
+            if (l.first == "AUTH") {
+                std::map<std::string,std::string> m = ParseTorReplyMapping(l.second);
+                std::map<std::string,std::string>::iterator i;
+                if ((i = m.find("METHODS")) != m.end())
+                    boost::split(methods, i->second, boost::is_any_of(","));
+                if ((i = m.find("COOKIEFILE")) != m.end())
+                    cookiefile = i->second;
```

Comment:
Correct. See section 3.21 of https://gitweb.torproject.org/torspec.git/tree/control-spec.txt (just constrained to be an absolute path). I don't think there's a vulnerability here, because all `zcashd` will do is open the file, read the first 32 bytes, and use that as a partial input to a hash function.